ssl/ssl_buffer.cc

```diff
@@ -121,6 +123,85 @@ void SSLBuffer::DiscardConsumed() {
   }
 }
 
+// An SSLBuffer is serialized as the following ASN.1 structure:
+//
+// SSLBuffer ::= SEQUENCE {
+//    version                           INTEGER (1),  -- SSLBuffer structure version
+//    bufAllocated                      BOOLEAN,
+//    offset                            INTEGER,
+//    size                              INTEGER,
+//    cap                               INTEGER,
+//    buf                               OCTET STRING,
+// }
+static const unsigned kSSLBufferVersion = 1;
+
+bool SSLBuffer::DoSerialization(CBB *cbb) {
+  if (!cbb) {
+    return false;
+  }
+  CBB seq;
+  if (!CBB_add_asn1(cbb, &seq, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1_uint64(&seq, kSSLBufferVersion) ||
+      !CBB_add_asn1_bool(&seq, (buf_allocated_ ? 1 : 0)) ||
+      !CBB_add_asn1_uint64(&seq, offset_) ||
+      !CBB_add_asn1_uint64(&seq, size_) ||
+      !CBB_add_asn1_uint64(&seq, cap_) ||
+      (buf_allocated_ && !CBB_add_asn1_octet_string(&seq, buf_, buf_size_)) ||
+      (!buf_allocated_ && !CBB_add_asn1_octet_string(&seq, inline_buf_, SSL3_RT_HEADER_LENGTH))) {
+    OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+    return false;
+  }
+  return CBB_flush(cbb) == 1;
+}
+
+bool SSLBuffer::DoDeserialization(CBS *cbs) {
+  if (!cbs) {
+    return false;
+  }
+
+  CBS seq, buf;
+  int buf_allocated_int;
+  uint64_t version, offset, size, cap;
+  if (!CBS_get_asn1(cbs, &seq, CBS_ASN1_SEQUENCE) ||
+      !CBS_get_asn1_uint64(&seq, &version) ||
+      version != kSSLBufferVersion ||
+      !CBS_get_asn1_bool(&seq, &buf_allocated_int) ||
+      !CBS_get_asn1_uint64(&seq, &offset) ||
+      !CBS_get_asn1_uint64(&seq, &size) ||
+      !CBS_get_asn1_uint64(&seq, &cap) ||
+      !CBS_get_asn1(&seq, &buf, CBS_ASN1_OCTETSTRING) ||
+      CBS_len(&seq) != 0) {
+    OPENSSL_PUT_ERROR(SSL, SSL_R_INVALID_SSL_BUFFER);
+    return false;
+  }
+  bool buf_allocated = !!buf_allocated_int;
+  if (buf_allocated) {
+    if (CBS_len(&buf) <= sizeof(inline_buf_)) {
+      OPENSSL_PUT_ERROR(SSL, SSL_R_INVALID_SSL_BUFFER);
+      return false;
+    }
+    buf_ = (uint8_t *)malloc(CBS_len(&buf));
```

Comment:
Hmm, the variable-length buffer is allocated in a fancy way in `SSLBuffer::EnsureCap()`; it is ensured that `buf_ + offset_ + header_len` (i.e the record body) is aligned to a `SSL3_ALIGN_PAYLOAD` boundary - this supposedly improves performance.

How is this ensured here?

Reply:
If you can argue it is not needed then that is fine. Looks negligible for a single packet to lose this optimisation.

Reply:
Discussed offline. The alignment may help some optimization. To ensure the property, it can increase some complexity of implementation. Without it, the functionality/correctness is not affected.

Reply:
I believe we have correctness. further, I believe these buffers are only used per-packet. So the process with the transferred session should quickly allocate new buffers which are aligned.